taichi/analysis/detect_for_with_break.cpp

```diff
@@ -0,0 +1,56 @@
+#include "taichi/ir/ir.h"
+#include <set>
+
+TLANG_NAMESPACE_BEGIN
+
+class DetectForWithBreak : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+
+  std::vector<Stmt *> loop_stack;
+  std::set<Stmt *> fors_with_break;
+  IRNode *root;
+
+  DetectForWithBreak(IRNode *root) : root(root) {
+  }
+
+  void visit(FrontendWhileStmt *stmt) override {
+    loop_stack.push_back(stmt);
+    stmt->body->accept(this);
+    loop_stack.pop_back();
+  }
+
+  void visit(FrontendForStmt *stmt) override {
+    loop_stack.push_back(stmt);
+    stmt->body->accept(this);
+    loop_stack.pop_back();
+  }
+
+  void visit(FrontendIfStmt *stmt) override {
```

Comment:
Nice implementation here! Maybe we should put this in `BasicStmtVisitor` in a future PR, something like `Make BasicStmtVisitor support frontend statements`.

Reply:
Potentially a boolean like `visit_container = true`?

Reply:
`BasicStmtVisitor` already means it visits all the containers. We just need to add support for frontend containers for it.

Reply:
I realized that we need to record the capturing loops here, so a general-purpose frontend statement visitor won't help much. You current impl looks good.